### DISCUSSION TITLE
Help with lmod updates: hardcoded versions

### DISCUSSION THREAD

**Discussion by mathomp4** [Homebrew/discussions#3306] [Everyday usage]
My guess is this is a "that's the way it is" but perhaps it's a "maybe I'm doing something wrong" so I thought I'd ask. 

To wit, yesterday I had lmod 8.7.2 and today I have 8.7.3 after a `brew upgrade`. Nice. But, lmod updates are always "fun" because all my shells from yesterday have, say,:

```
LMOD_ROOT=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.2/..
LMOD_PKG=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.2
LMOD_DIR=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.2/libexec
LMOD_CMD=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.2/libexec/lmod
```
but after the update when I try to run I get:
```
❯ module list
module:21: no such file or directory: /Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.2/libexec/lmod
```

(Yes, yes, I know. It's in `~/.homebrew` but this is a work laptop and I don't have rights to write to `/usr`)

This happens because the `module` command is a shell function:
```
❯ type module
module is a shell function from /Users/mathomp4/.homebrew/brew/opt/lmod/init/zsh
```
and in the init file has hardcoded paths to Cellar-with-versions. Yesterday it was 8.7.2 and now it's 8.7.3, and all the 8.7.2 paths are *poof*. So shell functions break, etc.

Now, the obviously easy solution is just "close and restart terminal", but, well, sometimes I like to keep around the shell since I have a nice workflow going on in the history and, well, I'm lazy.

So I guess a question is: is there perhaps a different way to "install" lmod so that it's agnostic to the version within?

**Comment by SMillerDev** [comment]
You can use `/Users/mathomp4/.homebrew/brew/opt/lmod/` in the variables.

**Reply by mathomp4** [reply]
@SMillerDev I agree, but the issue is that the `/Users/mathomp4/.homebrew/brew/opt/lmod/init/zsh` script itself seems to be installed via Brew with hardcoded paths:

```zsh
SUPPORT_KSH="no"
if [ $SUPPORT_KSH = yes -o -n "${KSH_VERSION+x}" ]; then
  if [ -z "${ZSH_VERSION+x}" -a -z "${__LMOD_SET_FPATH+x}" ]; then
    export FPATH=$(/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.3/libexec/addto FPATH /Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.3/init/ksh_funcs)
    export __LMOD_SET_FPATH=1
  fi
fi
unset SUPPORT_KSH

export LMOD_ROOT=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.3/..
export LMOD_PKG=/Users/mathomp4/.homebrew/brew/Cellar/lmod/8.7.3
export LMOD_DIR=$LMOD_PKG/libexec
export LMOD_CMD=$LMOD_PKG/libexec/lmod
export MODULESHOME=$LMOD_PKG
```

And so when you source this you get Cellar paths. It might just be the way things are. 🤷🏼

**Reply by SMillerDev** [reply]
If there's a configure option to change that you could make a pull request to fix it. But additionally, can't you just source the file again?

**Reply by danielnachun** [reply]
This may also reflect a large issue with `brew` not replacing cellar references with `opt` references in text files.  That's a more complex thing to fix but we remove cellar references from RPATHs and install names in binaries already so it would make sense to do this for text files as well.
